### PR TITLE
support threads in precompile and jl_task_wait_empty

### LINKED ISSUE #52435: Precompilation can't handle serialization of Task's enqued in Channels
Splitting this out into a separate issue.

From @Liozou 
> On https://github.com/JuliaLang/julia/pull/52405 I have a different crash in a different context: `]generate` a new package, for instance let's call it `PrecompileCrash`, and put the following in your `src/PrecompileCrash.jl`
```julia
module PrecompileCrash

const channel = Channel{Nothing}(0)
Base.Threads.@spawn take!($channel)

end # module PrecompileCrash
```
Then try to execute 
```julia -t1 --startup-file=no -e "using InteractiveUtils; versioninfo(); using PrecompileCrash"```

I get:
```
Julia Version 1.11.0-DEV.1026
Commit 96c4164145 (2023-12-05 18:18 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
  WORD_SIZE: 64
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
  Threads: 1 on 8 virtual cores
Environment:
  JULIA_NUM_THREADS = 4

[22573] signal 6 (-6): Aborted
in expression starting at none:1
pthread_kill at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
raise at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
get_item_for_reloc at /LionelSSDext4/liozou/julia/src/staticdata.c:1905 [inlined]
jl_read_reloclist at /LionelSSDext4/liozou/julia/src/staticdata.c:1981
jl_restore_system_image_from_stream_ at /LionelSSDext4/liozou/julia/src/staticdata.c:3159
jl_restore_package_image_from_stream at /LionelSSDext4/liozou/julia/src/staticdata.c:3580
jl_restore_incremental_from_buf at /LionelSSDext4/liozou/julia/src/staticdata.c:3627
ijl_restore_package_image_from_file at /LionelSSDext4/liozou/julia/src/staticdata.c:3715
_include_from_serialized at ./loading.jl:1073
_include_from_serialized at ./loading.jl:1048 [inlined]
_require_search_from_serialized at ./loading.jl:1626
_require at ./loading.jl:2039
__require_prelocked at ./loading.jl:1916
jfptr___require_prelocked_63238 at /LionelSSDext4/liozou/julia/usr/lib/julia/sys.so (unknown line)
jl_apply at /LionelSSDext4/liozou/julia/src/julia.h:2142 [inlined]
jl_f__call_in_world at /LionelSSDext4/liozou/julia/src/builtins.c:888
#invoke_in_world#3 at ./essentials.jl:989 [inlined]
invoke_in_world at ./essentials.jl:986 [inlined]
_require_prelocked at ./loading.jl:1907
macro expansion at ./loading.jl:1845 [inlined]
macro expansion at ./lock.jl:267 [inlined]
__require at ./loading.jl:1806
jfptr___require_63170 at /LionelSSDext4/liozou/julia/usr/lib/julia/sys.so (unknown line)
jl_apply at /LionelSSDext4/liozou/julia/src/julia.h:2142 [inlined]
jl_f__call_in_world at /LionelSSDext4/liozou/julia/src/builtins.c:888
#invoke_in_world#3 at ./essentials.jl:989 [inlined]
invoke_in_world at ./essentials.jl:986 [inlined]
require at ./loading.jl:1799
jfptr_require_63167 at /LionelSSDext4/liozou/julia/usr/lib/julia/sys.so (unknown line)
jl_apply at /LionelSSDext4/liozou/julia/src/julia.h:2142 [inlined]
call_require at /LionelSSDext4/liozou/julia/src/toplevel.c:484 [inlined]
eval_import_path at /LionelSSDext4/liozou/julia/src/toplevel.c:521
jl_toplevel_eval_flex at /LionelSSDext4/liozou/julia/src/toplevel.c:757
jl_toplevel_eval_flex at /LionelSSDext4/liozou/julia/src/toplevel.c:884
jl_toplevel_eval_flex at /LionelSSDext4/liozou/julia/src/toplevel.c:884
ijl_toplevel_eval_in at /LionelSSDext4/liozou/julia/src/toplevel.c:992
eval at ./boot.jl:428 [inlined]
exec_options at ./client.jl:291
_start at ./client.jl:525
jfptr__start_64763 at /LionelSSDext4/liozou/julia/usr/lib/julia/sys.so (unknown line)
jl_apply at /LionelSSDext4/liozou/julia/src/julia.h:2142 [inlined]
true_main at /LionelSSDext4/liozou/julia/src/jlapi.c:586
jl_repl_entrypoint at /LionelSSDext4/liozou/julia/src/jlapi.c:738
main at /LionelSSDext4/liozou/julia/cli/loader_exe.c:58
unknown function (ip: 0x7f3e53382d8f)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_start at /home/liozou/julia/usr/bin/julia (unknown line)
Allocations: 757403 (Pool: 757387; Big: 16); GC: 1
Aborted (core dumped)
```

_Originally posted by @Liozou in https://github.com/JuliaLang/julia/issues/52363#issuecomment-1843646320_

### PR DESCRIPTION
**Pull request by vtjnash**
Add a `nrunning` counter which identifies (when zero) when there is nothing running anymore. Allowing us to gate all tasks on all threads on reaching a quiescent state, not just thread 0. This should let us better support running precompile with threads (since we will be ensured that all of them are asleep in a consistent state before serialization tries to inspect the process state). We could additionally stop them afterwards to make sure there is no way for them to begin running, even if we forgot about some other event source, but that seems unnecessary paranoia for now.

Fixes #52435

### CODE REVIEW

File: src/jl_uv.c [hunk 1]
```diff
@@ -174,9 +179,12 @@ static void jl_uv_closeHandle(uv_handle_t *handle)
         ct->world_age = last_age;
         return;
     }
-    if (handle == (uv_handle_t*)&signal_async || handle == (uv_handle_t*)&wait_empty_worker)
+    if (handle == (uv_handle_t*)&wait_empty_worker)
+        handle->type = 0;
```
Comment (ararslan) [inline src/jl_uv.c:183]:
```suggestion
        handle->type = UV_UNKNOWN_HANDLE;
```
The implicit conversion here breaks building on musl Linux.

File: src/scheduler.c [hunk 2]
```diff
@@ -533,6 +579,23 @@ JL_DLLEXPORT jl_task_t *jl_task_get_next(jl_value_t *trypoptask, jl_value_t *q,
...
+void scheduler_delete_thread(jl_ptls_t ptls) JL_NOTSAFEPOINT
+{
+    if (jl_atomic_exchange_relaxed(&ptls->sleep_check_state, sleeping_like_the_dead) != sleeping) {
+        int wasrunning = jl_atomic_fetch_add_relaxed(&nrunning, -1);
+        if (wasrunning == 1) {
+            jl_ptls_t ptls2 = jl_atomic_load_relaxed(&jl_all_tls_states)[0];
+            // This was the last running thread, and there is no thread with !may_sleep
+            // so make sure tid 0 is notified to check wait_empty
+            uv_mutex_lock(&ptls2->sleep_lock);
+            uv_cond_wait(&ptls2->wake_signal, &ptls->sleep_lock);
```
Comment (vchuravy) [inline src/scheduler.c:591]:
```suggestion
            uv_cond_signal(&ptls2->wake_signal);
```

File: src/scheduler.c [hunk 3]
```diff
@@ -533,6 +579,23 @@ JL_DLLEXPORT jl_task_t *jl_task_get_next(jl_value_t *trypoptask, jl_value_t *q,
...
+            uv_mutex_unlock(&ptls2->sleep_lock);
```
Comment (vchuravy) [inline src/scheduler.c:592]:
```suggestion
            uv_cond_signal(&ptls2->wake_signal);
```

**Review by vchuravy — APPROVED** [review]

File: src/scheduler.c [hunk 4]
```diff
@@ -485,41 +506,64 @@ JL_DLLEXPORT jl_task_t *jl_task_get_next(jl_value_t *trypoptask, jl_value_t *q,
...
-                    }
+                task = wait_empty;
+                if (ptls->tid == 0 && task && jl_atomic_load_relaxed(&nrunning) == 0) {
+                    wasrunning = jl_atomic_fetch_add_relaxed(&nrunning, 1);
+                    assert(!wasrunning);
+                    wasrunning = !set_not_sleeping(ptls);
+                    assert(!wasrunning);
+                    JL_PROBE_RT_SLEEP_CHECK_TASK_WAKE(ptls);
+                    if (ct != task)
+                        jl_gc_enable_finalizers(ct, 0); // this annoyingly is rather sticky (we should reset it at the end of jl_task_wait_empty)
```
Comment (ararslan) [inline src/scheduler.c:555]: This is triggering an error in the GC checker:
```
error: Calling potential safepoint as SimpleFunctionCall from function annotated JL_NOTSAFEPOINT [julia.GCChecker]
```
Though I don't understand why.

**Comment by ufechner7** [comment]
Should this be back-ported to 1.10 ?

**Comment by vtjnash** [comment]
No, it is a new feature

**Comment by vtjnash** [comment]
Just the error for staticdata could be backported though (https://github.com/JuliaLang/julia/pull/52445/files#diff-0e5a5cdf744ddbc1c815bf0a767bc9988e37d88792e8acc95c7ed9863491f2af)

**Comment by IanButterworth** [comment]
@ufechner7 just checking you are handling the backport that Jameson suggested here

**Pull request merged**